drivers: build as static library

dcpleung opened the pull request:
Most of the driver types are using `zephyr_library()` and `zephyr_library_sources*()` already. This converts the remaining ones so driver builds are aligned. This avoids private include directories being spilled outside the driver build.

dcpleung committed:
drivers: timer: build as static library

Instead of putting object files inside libzephyr.a,
simply build a separate static library as most other
driver types are doing this already.

Also sort the entries alphabetically.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
drivers: console: build as static library

Instead of putting object files inside libzephyr.a,
simply build a separate static library as most other
driver types are doing this already.

Also sort the entries alphabetically.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
drivers: interrupt_controller: build as static library

Instead of putting object files inside libzephyr.a,
simply build a separate static library as most other
driver types are doing this already.

Also sort the entries alphabetically.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
drivers: i2c: slave: group into i2c static library

This changes the build command so the I2C EEPROM slave
is being grouped into the I2C static library.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
drivers: pcie: build as static library

Instead of putting the object files inside libzephyr.a,
simply build a separate static library as most other
driver types are doing this already.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
drivers: display: build as static library

Instead of putting object files inside libzephyr.a,
simply build a separate static library as most other
driver types are doing this already.

Also sort the entries alphabetically.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
drivers: bluetooth: build as static library

Instead of putting object files inside libzephyr.a,
simply build a separate static library as most other
driver types are doing this already.

Note that the include to subsys/bluetooth has been
moved from under drivers/bluetooth to subsys/bluetooth,
as it is actually the subsystem's job to make sure
the include directories are correct.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
drivers: wifi: build as static library

Instead of putting object files inside libzephyr.a,
simply build a separate static library as most other
driver types are doing this already.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
drivers: sensors: build as static libraries

Instead of putting object files inside libzephyr.a, simply
build a separate static library for the top level of sensor
drivers. Also, for those that were not building its own
static library, make them do so as majority of sensor
drivers are building their own static libraries.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
drivers: can: build as static library

Instead of putting object files inside libzephyr.a,
simply build a separate static library as most other
driver types are doing this already.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
drivers: clock_control: build as static library

Instead of putting object files inside libzephyr.a,
simply build a separate static library as most other
driver types are doing this already.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
drivers: disk: build as static library

Instead of putting object files inside libzephyr.a,
simply build a separate static library as most other
driver types are doing this already.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
drivers: edac: build as static library

Instead of putting object files inside libzephyr.a,
simply build a separate static library as most other
driver types are doing this already.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
drivers: ethernet: build as static library

Instead of putting object files inside libzephyr.a,
simply build a separate static library as most other
driver types are doing this already.

Note that the include to subsys has been moved from
under the drivers into subsys, as it is actually
the subsystem's job to make sure the include
directories are correct.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
drivers: flash: build as static library

Instead of putting object files inside libzephyr.a,
simply build a separate static library as most other
driver types are doing this already.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
drivers: hwinfo: build as static library

Instead of putting object files inside libzephyr.a,
simply build a separate static library as most other
driver types are doing this already.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
drivers: ieee802154: build as static library

Instead of putting object files inside libzephyr.a,
simply build a separate static library as most other
driver types are doing this already.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
drivers: led: build as static library

Instead of putting object files inside libzephyr.a,
simply build a separate static library as most other
driver types are doing this already.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
drivers: led_strip: build as static library

Instead of putting object files inside libzephyr.a,
simply build a separate static library as most other
driver types are doing this already.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
drivers: memc: build as static library

Instead of putting object files inside libzephyr.a,
simply build a separate static library as most other
driver types are doing this already.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
drivers: modem: build as static library

Instead of putting object files inside libzephyr.a,
simply build a separate static library as most other
driver types are doing this already.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
drivers: net: build as static library

Instead of putting object files inside libzephyr.a,
simply build a separate static library as most other
driver types are doing this already.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
drivers: pinmux: build as static library

Instead of putting object files inside libzephyr.a,
simply build a separate static library as most other
driver types are doing this already.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
drivers: ptp_clock: build as static library

Instead of putting object files inside libzephyr.a,
simply build a separate static library as most other
driver types are doing this already.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
drivers: usb: build as static library

Instead of putting object files inside libzephyr.a,
simply build a separate static library as most other
driver types are doing this already.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
drivers: watchdog: build as static library

Instead of putting object files inside libzephyr.a,
simply build a separate static library as most other
driver types are doing this already.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
cmake: add zephyr_library_include_directories_ifdef()

The zephyr_library_* variant is missing the ability to
include directories based on kconfig. So add the function.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
bluetooth: audio: build as static library

Make subsys/bluetooth/audio build as a static library.
This is in preparation to enable building the bluetooth
drivers as a static library. That would change the line
"zephyr_include_directories(subsys/bluetooth)" where
this include path is applied to all code compiled.
Its removal will result in subsys/bluetooth/audio not
able to find subsys/bluetooth/common/log.h. So change
subsys/bluetooth/audio to be built as a static library
so it can link to the subsys/bluetooth library where
it can find "log.h" there.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
drivers: sys_con: build as static library

Instead of putting object files inside libzephyr.a,
simply build a separate static library as most other
driver types are doing this already.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
drivers: pm_cpu_ops: build as static library

Instead of putting object files inside libzephyr.a,
simply build a separate static library as most other
driver types are doing this already.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
console: native_posix: change init priority to 99

This changes the init priority of native posix console to 99.
When building for native_posix, it is usually assumed that
the output would be stdout. With old priority at 60, UART
console, for example, can be initialized later which means
printk() would not go to stdout. So changing the native posix
console to be (hopefully) initialized last.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
display: change init priority config to 85

Some display drivers may actually be initialized after LittlevGL
as those drivers and LittlevGL's lvgl_init() all have SYS_INIT()
at APPLICATION and init priority the same as application init
priority. Depending on how the final binary is linked, these
drivers may initialize after lvgl_init() resulting in it not
able to find a display driver. This changes the value of
CONFIG_DISPLAY_INIT_PRIORITY so that the display drivers that
make use of this kconfig are initialized a bit earlier to
ensure that they are actually initialized before lvgl_init().

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
arm64: smp: arm64_smp_init to be done at PRE_KERNEL_2

The arm64_smp_init() is the same initialization level
and priority as the GICv3 interrupt controller. This means
that arm64_smp_init() can be called before the interrupt
controller driver has been initialized if linker decides
to put the driver init entry later. This would result in
faults when arm64_smp_init() is trying to connect interrupts.
So move arm64_smp_init() to PRE_KERNEL_2 instead. SMP
initialization is called later in the boot process so
this should not affect SMP operations.

This is in preparation of making interrupt controller
drivers to be build as static library. The linking order
is going to change which will result in this being
initialized before the interrupt contoller driver.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

gmarull commented on drivers/ieee802154/CMakeLists.txt:
```diff
@@ -1,20 +1,29 @@
+zephyr_library_sources_ifdef(CONFIG_IEEE802154_TELINK_B91	ieee802154_b91.c)
 
 if(CONFIG_SOC_CC1352R)
      target_compile_definitions(zephyr PRIVATE "DeviceFamily_CC13X2")
+     zephyr_library_compile_definitions("DeviceFamily_CC13X2")
```
duplicate now?

dcpleung replied:
Removed both redundant `target_compile_definitions()`.

gmarull commented on drivers/display/display_dummy.c:
```diff
@@ -126,5 +126,5 @@ static const struct display_driver_api dummy_display_api = {
 DEVICE_DEFINE(dummy_display, CONFIG_DUMMY_DISPLAY_DEV_NAME,
 		    &dummy_display_init, NULL,
 		    &dummy_display_data, NULL,
-		    APPLICATION, CONFIG_APPLICATION_INIT_PRIORITY,
```
we should probably introduce a display init priority together with this change, keeping `CONFIG_APPLICATION_INIT_PRIORITY` doesn't make much sense.

dcpleung replied:
I thought about it but decided not to change too many things at a time. I can submit another PR just for the display init priority. Or change add it in this PR if you want.

gmarull replied:
In general I feel that priority changes have no relation with the library changes (maybe I have missed something, though), so they may be part of another PR.

gmarull replied:
let's leave that for a 2nd PR then, I think it's fine.

gmarull commented on drivers/display/CMakeLists.txt:
```diff
@@ -1,25 +1,26 @@
+zephyr_library_sources_ifdef(CONFIG_ST7735R		display_st7735r.c)
 
 if (CONFIG_ILI9XXX)
-  zephyr_sources(display_ili9xxx.c)
+  zephyr_library_sources(display_ili9xxx.c)
 endif()
```
can be replaced with `zephyr_library_sources_ifdef`

gmarull reviewed (commented):
LGTM, just a few comments

dcpleung commented on drivers/console/semihost_console.c:
```diff
@@ -6,7 +6,6 @@
 #include <kernel.h>
 #include <device.h>
 #include <init.h>
-#include <kernel_arch_interface.h>
```
Yes. `kernel/include/kernel_arch_interface.h` is no longer in the include paths so it would result in built error. Luckily, nothing in that header file is being used, so it is removed.

gmarull replied:
can this be described in commit message?

dcpleung replied:
Added a separate commit for the removal.

dcpleung commented on drivers/i2c/slave/CMakeLists.txt:
```diff
@@ -1,3 +1,3 @@
 # SPDX-License-Identifier: Apache-2.0
 
-zephyr_sources_ifdef(CONFIG_I2C_EEPROM_SLAVE	eeprom_slave.c)
```
Since the upper level `drivers/i2c` has a `zephyr_library()`, the object file will be included in I2C library. I could add `zephyr_library()` here if you want.

gmarull replied:
ok, fair enough

dcpleung committed:
console: semihosting: remove unused include

Content in kernel_arch_interface.h is not being used.
So remove the #include.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

gmarull reviewed (approved):
Thanks for this cleanup

MaureenHelm commented on arch/arm64/core/smp.c:
```diff
@@ -237,6 +237,6 @@ static int arm64_smp_init(const struct device *dev)
 	return 0;
 }
-SYS_INIT(arm64_smp_init, PRE_KERNEL_1, CONFIG_KERNEL_INIT_PRIORITY_DEFAULT);
+SYS_INIT(arm64_smp_init, PRE_KERNEL_2, CONFIG_KERNEL_INIT_PRIORITY_DEFAULT);
```
What does this change have to do with building drivers as a static library?

dcpleung replied:
`arm64_smp_init()` has the same init level and priority as the GICv3 interrupt controller. The linking order is affected by the interrupt controller driver being inside a static library, resulting in the interrupt controller being initialized later than `arm64_smp_init()`. There are hard faults running `arm64_smp_init()` as it is setting up IRQs for IPI. This change is simply to make sure the interrupt controller is initialized first.

dcpleung replied:
This change is under a separate commit (https://github.com/zephyrproject-rtos/zephyr/pull/37512/commits/1c156742119e983af89b64fde5730e30b8752c30) with explanation in the commit message.

MaureenHelm replied:
The commit message doesn't mention why the link order changed

dcpleung replied:
Ah... because that commit is before the changes on driver. I have updated the commit message to say that it is in preparation for the driver changes. Does this work?

MaureenHelm replied:
Yes, thanks

MaureenHelm commented on drivers/console/Kconfig:
```diff
@@ -323,7 +323,7 @@ config XTENSA_CONSOLE_INIT_PRIORITY
 config NATIVE_POSIX_CONSOLE_INIT_PRIORITY
 	int "Init priority"
-	default 60
+	default 99
```
What does this change have to do with building drivers as a static library?

dcpleung replied:
This is due to linking order. For example, if both native POSIX console and UART console are enabled, the UART console driver takes over the printk hook which means there is nothing being printed to the terminal. Tests are failing due to empty outputs.

dcpleung commented:
Rebased and updated after the display init priority PR was merged. Instead of changing the init level of the display driver, change the default on the init priority kconfig so the display drivers are initialized a bit earlier.

dcpleung commented:
The bluetooth test failure was traced to commit 27b8beaa22db29f2d880cbe674a22eeb8418d91d. Note that the test seems to pass but reported failure:

```
d_00: @00:00:23.291564 INFO: AdvX tests Passed
d_01: @00:00:23.283638  main: The TESTCASE FAILED with return code 1
```

alexanderwachter reviewed (approved):
+1 for CAN
